pennylane/beta/tapes/tape.py

```diff
@@ -41,6 +42,312 @@
 )
 
 
```

Comment:
Best to read the following functions from the last to the first. They contain some hectic recursion, but hope the docstrings help.